numpy/lib/shape_base.py

```diff
@@ -89,9 +90,20 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     outshape = asarray(arr.shape).take(indlist)
     i.put(indlist, ind)
     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-    #  if res is a number, then we have a smaller output array
+
+    # see gh-8363
+    _strcase = False
+    _dtypes = set()
+
+    # If res is a number, then we have a smaller output array.
     if isscalar(res):
```

Comment:
@gfyoung: you're correct, it would fail in the array case. I haven't run your tests against it, because I don't really want to mix separate features requests into one PR.